utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +697,395 @@ impl FromStr for FixedDecimal {
+    /// assert_eq!(decimal.writeable_to_string(), "0.012345678");
+    ///
+    /// let decimal = FixedDecimal::new_from_f64(-123456.78, DoublePrecision::Maximum).unwrap();
+    /// assert_eq!(decimal.writeable_to_string(), "-123456.78");
+    ///
+    /// let decimal = FixedDecimal::new_from_f64(12345678000., DoublePrecision::Maximum).unwrap();
+    /// assert_eq!(decimal.writeable_to_string(), "12345678000.0");
+    /// ```
+    pub fn new_from_f64(float: f64, precision: DoublePrecision) -> Result<Self, Error> {
+        let mut decimal = Self::new_from_f64_raw(float)?;
+        match precision {
+            DoublePrecision::Maximum => (),
+            DoublePrecision::Magnitude(mag) => {
+                let n_digits = decimal.digits.len() as i16;
+                // magnitude of the lowest digit in self.digits
+                let lowest_magnitude = decimal.magnitude - n_digits + 1;
+
+                if mag > lowest_magnitude {
+                    let round_by = (mag - lowest_magnitude) as i16;
+
+                    if round_by <= n_digits {
+                        decimal.round_digits(round_by as u16);
+                    } else {
+                        // If we need to round by more digits than rounding can ever produce
+                        // the number is zero
+                        decimal = Default::default();
+                    }
+                }
+                if mag < 0 {
+                    // If the target magnitude was negative, make
+                    // sure we update the lower magnitude to match it
+                    decimal.lower_magnitude = mag;
+                } else if decimal.lower_magnitude < 0 {
+                    // If the target magnitude was positive,
+                    // we may have truncated digits off the right
+                    // side, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+            DoublePrecision::SignificantDigits(sig) => {
+                let sig = sig as i16;
+                if sig == 0 {
+                    return Err(Error::Limit);
+                }
+                let n_digits = decimal.digits.len() as i16;
+
+                if sig < n_digits {
+                    let round_by = (n_digits - sig) as i16;
+                    decimal.round_digits(round_by as u16);
+                    debug_assert!(decimal.digits.len() >= sig as usize);
+                    if decimal.digits.len() < sig as usize {
+                        // it's possible the magnitude increased by 1 due to overflow. round it again
+                        decimal.round_digits(1);
+                        // In case of overflow it cannot overflow again due to rounding
+                        debug_assert!(decimal.digits.len() == sig as usize);
+                    }
+                }
+                let target_lowest_magnitude = decimal.magnitude - sig as i16 + 1;
+                if target_lowest_magnitude <= 0 {
```

Comment:
This stuff should probably also be in `round_digits`.  I note that you need it both here and in `Magnitude`.